### ISSUE TITLE
[Bug]  Non-Cookie Site Data not cleared for Greylisted website on restart

### ISSUE THREAD

**Issue by LaurentGrenet** [mrdokenny/Cookie-AutoDelete#1199]
### Acknowledgements

- [X] I acknowledge that I have read the above items

### Describe the bug

LocalStorage of "greylisted" websites is not cleared at browser restart (and maybe also some other piece of data)
I detected the bug in particular with website https://www.lefigaro.fr/

The first time you access to this website, it asks for your consent to cookies, and records your answer, in order to do not ask you the same at your next access.
Your consent is not stored in a cookie (lefigaro.fr do not actually create any cookie) but in LocalStorage.

If there is no rule defined in CAD relative to this website, cleansing of this data (your consent) is correctlty done by CAD if you close the tab, a/o switch in the tab to another domain a/o restart the browser, and when the cleansing is done, re-opening the site asks again for your consent.

But if there is a "GreyList" rule defined on the domain, with all boxes checked ("Keep ... until restart"), the cleansing is not done if you "only" close the tab or switch to another domain (which is expected behaviour), but no cleansing as well is done EVEN IF YOU RESTART browser... whch is a bug.

### To Reproduce

Steps to reproduce 

Step 1:
To start on a "blank page", clear in Firefox all data relative to any website.
Do not create any rule in CAD, neither Blank nor Grey, relative to lefigaro.fr
Suspend automatic cleansing in CAD
Open www.lefigaro.fr and give (or reject) your consent to cookies
Close the tab, or switch to another domain, or restart the browser
Open again www.lefigaro.fr and see that your consent is no more asked
Activate automatic cleansing in CAD
Close the tab (the one opend on lefigaro.fr), or switch to another domain, or restart the browser
Wait for automatic cleansing
Open again www.lefigaro.fr and see that now your consent is again asked.

Step 2 :
To start also this 2nd step on a "blank page", clear again in Firefox all data relative to any website.
Create in CAD a "GreyList" rule, on "*.lefigaro.fr", with all boxes ("keep until restart") checked.
Activate automatic cleansing in CAD
Open www.lefigaro.fr and give (or reject) your consent to cookies
Close the tab, or switch to another domain.
Open again www.lefigaro.fr and verify that your consent is no more asked (normal, since "GreyList")
Restart the browser
Wait for automatic cleansing
Open again www.lefigaro.fr and see that your consent is still no more asked.... while it should if cleansing at restart has been effective.

### Expected Behavior

To have ALL data relative to GreyListed website (including the LocaStorage) cleared at browser restart,and then to have the website asking again for your consent to cookies.

### System Info - Operating System (OS)

Windows 10 Pro 21H1

### System Info - Browser Info

Firefox 90.0.2 (64 bits)

### System Info - CookieAutoDelete Version

3.6.0

**Comment by kennethtran93** [comment]
I've tried some initial testing, but will need more configuration details.  
- Firefox 'Restore previous session' setting

For testing purposes I presume the following CAD settings are enabled/checked.  Please confirm
  - Enable Greylist Cleanup on Browser Restart
  - Enable LocalStorage Cleanup

I've tried testing this on Firefox Dev Edition 91.0b9, though I'm not seeing anything similar to a cookie preference prompt (I also don't speak/read French so I cannot navigate around to know where to click to get the configuration set.  This is the view I see after completely clearing all site data, and turning off CAD autoclean.  The only other webextensions I have enabled are: Dark Reader, Cookie Manager, and Cookie Quick Manager
![image](https://user-images.githubusercontent.com/6724477/128279689-f4932419-8c67-4deb-8a5f-c3498bc1dd0b.png)

I do, however, see that there is something stored in localstorage akin to a 'start session timestamp', so I based off that for preliminary testing.  Also cookies are being stored by that website as well.

**Comment by kennethtran93** [comment]
In the event you have 'Restore Previous Session' checked in Firefox, do you have 'Clean Cookies and other Site Data from Open Tabs on Startup' also checked in CAD settings as well?

**Comment by kennethtran93** [comment]
Apologies - you may ignore my previous post.  Aside from not seeing the cookie preference I was able to confirm the bug from other localstorage data.

Thanks for bringing this up.

**Comment by LaurentGrenet** [comment]
Hi Kenneth,

even if you don't need them anymore, pls find below my answers to your questions 

- Firefox 'Restore previous session' setting : Off (not checked)
- Enable Greylist Cleanup on Browser Restart : On (checked)
- Enable LocalStorage Cleanup : On (checked)

I'm afraid that you are not able to see the cookies preference screen on lefigaro.fr not because you do not speak french, but because this screen maybe appears **only** to users residing in France (or in any country of European Union) : to get this content is indeed a legal obligation in UE (and in particular in France), maybe not in your country. Otherwise, you would have seen it, even not reading french, since it automatically displays on the screen (not in a separate pop up window) and you cannot miss it....

But if you've been able to reproduce the bug with another website, all's good....

**Comment by vpreyer** [comment]
Same thing happening to me. 
- FF 90.0.2
- CAD 3.6.0
- W10 21H1 19043.1083
- Automatic Cleaning Options: ALL checkboxes enabled
- LocalStorage Cleanup enabled
- wildcard (*) added to greylist

After a full cleanup (I have deleted all cookies through FF settings) I visited a few random sites, then followed these steps:
- closed the tabs
- restarted the browser
- hit the yellow Clean button (with no tabs open)
- checked stored cookies

In addition to whitelisted items, the cookies from the non-whitelisted websites are also kept, regardless of the settings. As for the previous comments, I don't think this is supposed to be site-specific. If it's not whitelisted it's supposed to go.

I have plenty of other privacy extensions installed, I don't know if it matters at all but it probably shouldn't.

**Comment by kennethtran93** [comment]
@vpreyer as you have wildcard for greylist, this means that any non-whitelisted item will match the greylist entry, thus the clean button will not do anything for you as those are supposedly cleaned at browser startup.  Cookies are still cleaned at startup, just not any other browsing data is the main issue being raised here.

**Comment by vpreyer** [comment]
@kennethtran93 Thanks for the explanation. I added the wildcard as a troubleshooting step after I realized cookies were not deleted and I read up on the issue. For a long time I didn't bother to fiddle with the settings, I only whitelisted some sites.

It is now removed and unwanted cookies get deleted on browser startup so please consider it resolved.

**Comment by bitpixl** [comment]
I'm also running into this issue.

- firefox 91.0.1-esr / clean profile
- W10 21H1
- CAD 3.6.0
- Automatic Cleaning Options: All checked
- Other Browsing Data Cleanup Options: All checked

easiest way to test that I found: 

- go to https://www.twitch.tv/
- greylist twitch.tv and only select "Keep LocalStorage until restart"
- click on the "little person icon" (next to sign up) on top right of twitch site
- deselect 'Dark theme' from dropdown menu (my clean profile gives dark mode by default)
- close twitch tab
- select clean from CAD (to be sure)
- restart browser (to be sure)
- go to https://www.twitch.tv/
- site is still set to the light theme

**Comment by m-rz** [comment]
Same issue for me. I believe that all greylisted websites are affected in my case. Found an online script that allows to reproduce the issue.

My setup:

- Win 10 Home 21H1
- CAD 3.6.0
- Firefox 92.0.1 
- Enable Automatic Cleaning: Yes
- Enable Greylist Cleanup on Browser Restart: Yes
- Clean All Expired Cookies: Yes
- Other Browsing Data Cleanup Options: enabled for cache, IndexDB and local storage
- Rules
  - A few domains whitelisted
  - Single greylist entry for `*`

Reproducing the issue:
- go to http://www.sharonminsuk.com/code/storage-test.html (caution, works properly only with HTTP)
- store a value inthe local storage: test_092021
- the value is stored in the local storage under key `theirValue` for host `http://www.sharonminsuk.com` as indicated by developer tools
- restart the browser
- go to http://www.sharonminsuk.com/code/storage-test.html again
- the value stored in the local storage is still present and is displayed by the script

Expected behavior:
- I assume that after restarting the browser the `theirValue:test_092021` entry in the local storage for `http://www.sharonminsuk.com` should be gone

Additional logs (filtered by the target domain):
```
CAD_3.6.0 - debug - CleanupService.filterSiteData: debug data.
{
  "notProtectedByOpenTab": true,
  "notInAnyLists": false,
  "siteData": "IndexedDB",
  "canCleanSiteData": false,
  "nonBlankCookieHostName": true,
  "notOpenTabAndCanClean": false,
  "CleanReasonObject": {
    "cached": false,
    "cleanCookie": true,
    "cookie": {
      "name": "CookieAutoDeleteBrowsingDataCleanup",
      "value": "***",
      "domain": "www.sharonminsuk.com",
      "hostOnly": true,
      "path": "/IIQ2CbBw_",
      "secure": false,
      "httpOnly": false,
      "sameSite": "no_restriction",
      "session": false,
      "firstPartyDomain": "",
      "expirationDate": 1664315900,
      "storeId": "firefox-default",
      "hostname": "sharonminsuk.com",
      "mainDomain": "sharonminsuk.com",
      "preparedCookieDomain": "http://www.sharonminsuk.com/IIQ2CbBw_"
    },
    "expression": {
      "expression": "*",
      "listType": "GREY",
      "storeId": "default",
      "cleanAllCookies": true,
      "cleanSiteData": [],
      "cookieNames": [],
      "id": "twadeiw4h"
    },
    "openTabStatus": "reasonTabsWereNotIgnored",
    "reason": "reasonCADSiteDataCookie"
  }
} common-background-popup-setting.bundle.js:2:43851
CAD_3.6.0 - debug - CleanupService.filterSiteData: IndexedDB cleanup returned false for sharonminsuk.com
common-background-popup-setting.bundle.js:2:43851
CAD_3.6.0 - debug - CleanupService.filterSiteData: debug data.
{
  "notProtectedByOpenTab": true,
  "notInAnyLists": false,
  "siteData": "LocalStorage",
  "canCleanSiteData": false,
  "nonBlankCookieHostName": true,
  "notOpenTabAndCanClean": false,
  "CleanReasonObject": {
    "cached": false,
    "cleanCookie": true,
    "cookie": {
      "name": "CookieAutoDeleteBrowsingDataCleanup",
      "value": "***",
      "domain": "www.sharonminsuk.com",
      "hostOnly": true,
      "path": "/IIQ2CbBw_",
      "secure": false,
      "httpOnly": false,
      "sameSite": "no_restriction",
      "session": false,
      "firstPartyDomain": "",
      "expirationDate": 1664315900,
      "storeId": "firefox-default",
      "hostname": "sharonminsuk.com",
      "mainDomain": "sharonminsuk.com",
      "preparedCookieDomain": "http://www.sharonminsuk.com/IIQ2CbBw_"
    },
    "expression": {
      "expression": "*",
      "listType": "GREY",
      "storeId": "default",
      "cleanAllCookies": true,
      "cleanSiteData": [],
      "cookieNames": [],
      "id": "twadeiw4h"
    },
    "openTabStatus": "reasonTabsWereNotIgnored",
    "reason": "reasonCADSiteDataCookie"
  }
} common-background-popup-setting.bundle.js:2:43851
CAD_3.6.0 - debug - CleanupService.filterSiteData: LocalStorage cleanup returned false for sharonminsuk.com
common-background-popup-setting.bundle.js:2:43851
CAD_3.6.0 - debug - TabEvents.onDomainChange: First mainDomain set.
{
  "tabId": 1,
  "changeInfo": {
    "status": "complete"
  },
  "mainDomain": "sharonminsuk.com",
  "partialTabInfo": {
    "cookieStoreId": "firefox-default",
    "discarded": false,
    "id": 1,
    "incognito": false,
    "status": "complete",
    "url": "http://www.sharonminsuk.com/code/storage-test.html",
    "windowId": 1
  }
} common-background-popup-setting.bundle.js:2:43851
CAD_3.6.0 - debug - TabEvents.onTabUpdate: action delay has been set for ~750 ms.
{
  "tabId": 1,
  "changeInfo": {
    "status": "complete"
  },
  "partialTabInfo": {
    "cookieStoreId": "firefox-default",
    "discarded": false,
    "id": 1,
    "incognito": false,
    "status": "complete",
    "url": "http://www.sharonminsuk.com/code/storage-test.html",
    "windowId": 1
  }
} common-background-popup-setting.bundle.js:2:43851
CAD_3.6.0 - debug - TabEvents.onTabUpdate: actions will now commence.
{
  "tabId": 1,
  "changeInfo": {
    "status": "complete"
  },
  "partialTabInfo": {
    "cookieStoreId": "firefox-default",
    "discarded": false,
    "id": 1,
    "incognito": false,
    "status": "complete",
    "url": "http://www.sharonminsuk.com/code/storage-test.html",
    "windowId": 1
  }
} common-background-popup-setting.bundle.js:2:43851
CAD_3.6.0 - debug - Libs.getAllCookiesForDomain:  browser.cookies.getAll for domain.
{
  "partialTabInfo": {
    "cookieStoreId": "firefox-default",
    "discarded": false,
    "id": 1,
    "incognito": false,
    "status": "complete",
    "url": "http://www.sharonminsuk.com/code/storage-test.html",
    "windowId": 1
  },
  "domain": "sharonminsuk.com"
} common-background-popup-setting.bundle.js:2:43851
CAD_3.6.0 - debug - Libs.getAllCookiesForDomain:  Filtered Cookie Count
{
  "partialTabInfo": {
    "cookieStoreId": "firefox-default",
    "discarded": false,
    "id": 1,
    "incognito": false,
    "status": "complete",
    "url": "http://www.sharonminsuk.com/code/storage-test.html",
    "windowId": 1
  },
  "tabURL": "http://www.sharonminsuk.com/code/storage-test.html",
  "hostname": "sharonminsuk.com",
  "cookieCount": 0
} common-background-popup-setting.bundle.js:2:43851
CAD_3.6.0 - debug - TabEvents.getAllCookieActions:  A temporary cookie has been set for future BrowsingData cleaning as the site did not set any cookies yet.
{
  "partialTabInfo": {
    "cookieStoreId": "firefox-default",
    "discarded": false,
    "id": 1,
    "incognito": false,
    "status": "complete",
    "url": "http://www.sharonminsuk.com/code/storage-test.html",
    "windowId": 1
  },
  "cadLSCookie": {
    "expirationDate": 1664316342,
    "firstPartyDomain": "",
    "name": "CookieAutoDeleteBrowsingDataCleanup",
    "path": "/P5mC5Qf0b",
    "storeId": "firefox-default",
    "url": "http://www.sharonminsuk.com/code/storage-test.html",
    "value": "CookieAutoDeleteBrowsingDataCleanup"
  }
} common-background-popup-setting.bundle.js:2:43851
CAD_3.6.0 - debug - TabEvents.onTabUpdate: action delay has been set for ~750 ms.
{
  "tabId": 1,
  "changeInfo": {
    "cookieChanged": {
      "removed": false,
      "cookie": {
        "name": "CookieAutoDeleteBrowsingDataCleanup",
        "value": "***",
        "domain": "www.sharonminsuk.com",
        "hostOnly": true,
        "path": "/P5mC5Qf0b",
        "secure": false,
        "httpOnly": false,
        "sameSite": "no_restriction",
        "session": false,
        "firstPartyDomain": "",
        "expirationDate": 1664316342,
        "storeId": "firefox-default"
      },
      "cause": "explicit"
    }
  },
  "partialTabInfo": {
    "cookieStoreId": "firefox-default",
    "discarded": false,
    "id": 1,
    "incognito": false,
    "status": "complete",
    "url": "http://www.sharonminsuk.com/code/storage-test.html",
    "windowId": 1
  }
} common-background-popup-setting.bundle.js:2:43851
CAD_3.6.0 - debug - TabEvents.onTabUpdate: actions will now commence.
{
  "tabId": 1,
  "changeInfo": {
    "cookieChanged": {
      "removed": false,
      "cookie": {
        "name": "CookieAutoDeleteBrowsingDataCleanup",
        "value": "***",
        "domain": "www.sharonminsuk.com",
        "hostOnly": true,
        "path": "/P5mC5Qf0b",
        "secure": false,
        "httpOnly": false,
        "sameSite": "no_restriction",
        "session": false,
        "firstPartyDomain": "",
        "expirationDate": 1664316342,
        "storeId": "firefox-default"
      },
      "cause": "explicit"
    }
  },
  "partialTabInfo": {
    "cookieStoreId": "firefox-default",
    "discarded": false,
    "id": 1,
    "incognito": false,
    "status": "complete",
    "url": "http://www.sharonminsuk.com/code/storage-test.html",
    "windowId": 1
  }
} common-background-popup-setting.bundle.js:2:43851
CAD_3.6.0 - debug - Libs.getAllCookiesForDomain:  browser.cookies.getAll for domain.
{
  "partialTabInfo": {
    "cookieStoreId": "firefox-default",
    "discarded": false,
    "id": 1,
    "incognito": false,
    "status": "complete",
    "url": "http://www.sharonminsuk.com/code/storage-test.html",
    "windowId": 1
  },
  "domain": "sharonminsuk.com"
} common-background-popup-setting.bundle.js:2:43851
CAD_3.6.0 - debug - Libs.getAllCookiesForDomain:  Filtered Cookie Count
{
  "partialTabInfo": {
    "cookieStoreId": "firefox-default",
    "discarded": false,
    "id": 1,
    "incognito": false,
    "status": "complete",
    "url": "http://www.sharonminsuk.com/code/storage-test.html",
    "windowId": 1
  },
  "tabURL": "http://www.sharonminsuk.com/code/storage-test.html",
  "hostname": "sharonminsuk.com",
  "cookieCount": 1
} common-background-popup-setting.bundle.js:2:43851
```

**Comment by redactedscribe** [comment]
Bug still preset: with a site greylisted and with only "Keep LocalStorage" checked for it, the site's local storage keys will be present and contain the same content after having restarted the browser. Hoping this will get fixed soon.

Cookie AutoDelete 3.6.0
Firefox 95.0.2

**Comment by kennethtran93** [comment]
This is included in 3.7.0 and has been released on Firefox.
